### ISSUE TITLE
Python 3.11-14 deprecations

### ISSUE THREAD

**Issue by nsoranzo** [galaxyproject/galaxy#16854]
List of deprecated stuff we are currently using:

- [x] [PEP 594](https://peps.python.org/pep-0594/) led to the deprecations of the following modules slated for removal in Python 3.13: [imghdr](https://docs.python.org/3/library/imghdr.html#module-imghdr)
- [ ] [datetime.datetime](https://docs.python.org/3/library/datetime.html#datetime.datetime)’s [utcnow()](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow) and [utcfromtimestamp()](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp) are deprecated and will be removed in a future version. Instead, use timezone-aware objects to represent datetimes in UTC: respectively, call [now()](https://docs.python.org/3/library/datetime.html#datetime.datetime.now) and [fromtimestamp()](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp) with the tz parameter set to [datetime.UTC](https://docs.python.org/3/library/datetime.html#datetime.UTC).
- [x] Deprecated the `importlib.abc.Traversable` class, scheduled for removal in Python 3.14. Use [importlib.resources.abc.Traversable](https://docs.python.org/3/library/importlib.resources.abc.html#importlib.resources.abc.Traversable) class instead.
- [x] The [lib2to3](https://docs.python.org/3/library/2to3.html#module-lib2to3) package and [2to3](https://docs.python.org/3/library/2to3.html#to3-reference) tool are now deprecated and slated for removal in Python 3.13. Also they may not be able to parse Python 3.10 or newer. See [PEP 617](https://peps.python.org/pep-0617/), introducing the new PEG parser, for details.
- [x] [tarfile](https://docs.python.org/3/library/tarfile.html#module-tarfile): Extracting tar archives without specifying filter is deprecated until Python 3.14, when 'data' filter will become the default. See [Extraction filters](https://docs.python.org/3/library/tarfile.html#tarfile-extraction-filter) for details. - Fixed in https://github.com/galaxyproject/galaxy/pull/19406 .
- [x] The [argparse.FileType](https://docs.python.org/3.14/library/argparse.html#argparse.FileType) type converter is deprecated in Python 3.14. Anything with resource management should be done downstream after the arguments are parsed. Fixed in https://github.com/galaxyproject/galaxy/pull/20876 .
- [x] The [codecs.open](https://docs.python.org/3.14/library/codecs.html#codecs.open) function has been deprecated since Python 3.14, use `open()` instead. Fixed by https://github.com/galaxyproject/galaxy/pull/20891 .

Others we may be using:
- [x] Remove the `filename` attribute of [gzip](https://docs.python.org/3/library/gzip.html#module-gzip)’s [gzip.GzipFile](https://docs.python.org/3/library/gzip.html#gzip.GzipFile), deprecated since Python 2.6, use the [name](https://docs.python.org/3/library/gzip.html#gzip.GzipFile.name) attribute instead. In write mode, the `filename` attribute added '.gz' file extension if it was not present.

More are listed at https://docs.python.org/3/whatsnew/3.11.html#deprecated , https://docs.python.org/3/whatsnew/3.12.html#deprecated , https://docs.python.org/3/whatsnew/3.13.html#removed-modules-and-apis and https://docs.python.org/3/whatsnew/3.13.html#new-deprecations https://docs.python.org/3.14/whatsnew/3.14.html#deprecated

**Comment by cat-bro** [comment]
The version of tensorflow in optional dependencies is too low to be compatible with python 3.11

**Comment by nsoranzo** [comment]
@cat-bro This is now tracked separately at https://github.com/galaxyproject/galaxy/issues/17857 .

**Comment by elichad** [comment]
For Python 3.12 you may already need to `apt install python3-lib2to3` depending on your distribution. I'm using 3.12 from the deadsnakes PPA and needed this step

**Comment by guillaume-kermorgant** [comment]
The version of tensorflow is still too low to be compatible with Python 3.13, should I open a separate issue for that as well?

**Comment by nsoranzo** [comment]
> The version of tensorflow is still too low to be compatible with Python 3.13, should I open a separate issue for that as well?

Yes, please! It looks like the minimum tensorflow version for Python 3.13 is 2.20.0 .

**Comment by guillaume-kermorgant** [comment]
Ok thanks, just opened https://github.com/galaxyproject/galaxy/issues/23373
